app/services/foreman/client_certificate.rb

```diff
@@ -0,0 +1,63 @@
+module Foreman
+  class ClientCertificate
+    delegate :logger, to: :Rails
+    attr_reader :request
+
+    def initialize(request:)
+      @request = request
+    end
+
+    def raw_cert_available?
+      @raw_cert_available ||= request.ssl? && request.env.has_key?(certificate_env_key) && raw_data.present? && raw_data != '(null)'
+    end
+
+    def raw_data
+      request.env[certificate_env_key]
+    end
+
+    def subject
+      return certificate_extract.subject if raw_cert_available?
+      dn = request.env[dn_key]
+      return unless dn && dn =~ /CN=([^\s\/,]+)/i
+
+      Regexp.last_match(1).downcase
+    end
+
+    def subject_alternative_names
+      return unless raw_cert_available?
+      certificate_extract.subject_alternative_names
+    end
+
+    def verified?
+      verify = request.env[verify_key]
+      return true if verify == 'SUCCESS'
+      logger.info { "Client certificate is invalid: #{verify}" }
+      false
+    end
+
+    def hosts
+      hosts_from_certificate = []
+      hosts_from_certificate << subject if subject.present?
```

Comment:
Not sure if the order is important, but if it is than in the previous implementation the subject was added after the alt names and now it is added before.

Reply:
Oh, good catch. Let's restore the previous behavior.